File: backend/internal/model/contest_orchestration.go
package model

import (
	"database/sql/driver"
	"encoding/json"
)

// ChallengeWorkspaceSpec 定义解题赛题目的主工作区。
type ChallengeWorkspaceSpec struct {
	Image            string            `json:"image"`
	DisplayName      string            `json:"display_name"`
	Template         string            `json:"template"`
	InteractionTools []string          `json:"interaction_tools"`
	Resources        map[string]string `json:"resources"`
	InitScripts      []string          `json:"init_scripts"`
}

// ChallengeServicePort 定义题目附加服务的端口暴露方式。
type ChallengeServicePort struct {
	Name     string `json:"name"`
	Port     int    `json:"port"`
	Protocol string `json:"protocol"`
	ExposeAs string `json:"expose_as"`
}

// ChallengeServiceSpec 定义解题赛题目的附加服务组件。
type ChallengeServiceSpec struct {
	Key         string                 `json:"key"`
	Image       string                 `json:"image"`
	Purpose     string                 `json:"purpose"`
	Description string                 `json:"description"`
	Ports       []ChallengeServicePort `json:"ports"`
	Env         JSONMap                `json:"env"`
}

// ChallengeTopologySpec 描述题目环境的拓扑和暴露方式。
type ChallengeTopologySpec struct {
	Mode          string   `json:"mode"`
	ExposedEntrys []string `json:"exposed_entries"`
	SharedNetwork bool     `json:"shared_network"`
}

// ChallengeForkSpec 描述题目环境的 Fork 复现配置。
type ChallengeForkSpec struct {
	Enabled      bool   `json:"enabled"`
	Chain        string `json:"chain"`
	ChainID      int    `json:"chain_id"`
	Label        string `json:"label"`
	RPCURL       string `json:"rpc_url"`
	BlockNumber  uint64 `json:"block_number"`
	TargetTxHash string `json:"target_tx_hash"`
}

// ChallengeScenarioSpec 描述真实漏洞题的复现场景。
type ChallengeScenarioSpec struct {
	ContractAddress string   `json:"contract_address"`
	AttackGoal      string   `json:"attack_goal"`
	InitSteps       []string `json:"init_steps"`
	SolveSteps      []string `json:"solve_steps"`
	DefenseGoal     string   `json:"defense_goal"`
}

// ChallengeLifecycleSpec 描述题目环境的生命周期策略。
type ChallengeLifecycleSpec struct {
	TimeLimitMinutes int  `json:"time_limit_minutes"`
	AutoDestroy      bool `json:"auto_destroy"`
	ReuseRunningEnv  bool `json:"reuse_running_env"`
}

// ChallengeValidationSpec 描述环境题的验证方式。
type ChallengeValidationSpec struct {
	Mode        string `json:"mode"`
	Description string `json:"description"`
}

// ChallengeOrchestration 定义解题赛题目的完整环境编排。
type ChallengeOrchestration struct {
	Mode             string                  `json:"mode"`
	NeedsEnvironment bool                    `json:"needs_environment"`
	Workspace        ChallengeWorkspaceSpec  `json:"workspace"`
	Services         []ChallengeServiceSpec  `json:"services"`
	Topology         ChallengeTopologySpec   `json:"topology"`
	Fork             ChallengeForkSpec       `json:"fork"`
	Scenario         ChallengeScenarioSpec   `json:"scenario"`
	Lifecycle        ChallengeLifecycleSpec  `json:"lifecycle"`
	Validation       ChallengeValidationSpec `json:"validation"`
}

// Scan 实现 Scanner 接口。
func (o *ChallengeOrchestration) Scan(value interface{}) error {
	if value == nil {
		*o = ChallengeOrchestration{}
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}

	return json.Unmarshal(bytes, o)
}

// Value 实现 Valuer 接口。
func (o ChallengeOrchestration) Value() (driver.Value, error) {
	return json.Marshal(o)
}

// BattleSharedChainSpec 描述对抗赛的共享链环境。
type BattleSharedChainSpec struct {
	Image           string            `json:"image"`
	ChainType       string            `json:"chain_type"`
	NetworkID       int               `json:"network_id"`
	BlockTime       int               `json:"block_time"`
	InitialBalances map[string]string `json:"initial_balances"`
	RPCURL          string            `json:"rpc_url"`
}

// BattleJudgeSpec 描述对抗赛的裁判与评分组件。
type BattleJudgeSpec struct {
	Image             string         `json:"image"`
	JudgeContract     string         `json:"judge_contract"`
	TokenContract     string         `json:"token_contract"`
	StrategyInterface string         `json:"strategy_interface"`
	ResourceModel     string         `json:"resource_model"`
	ScoringModel      string         `json:"scoring_model"`
	ScoreWeights      map[string]int `json:"score_weights"`
	AllowedActions    []string       `json:"allowed_actions"`
	ForbiddenCalls    []string       `json:"forbidden_calls"`
}

// BattleTeamWorkspaceSpec 描述每支队伍的开发环境模板。
type BattleTeamWorkspaceSpec struct {
	Image            string            `json:"image"`
	DisplayName      string            `json:"display_name"`
	InteractionTools []string          `json:"interaction_tools"`
	Resources        map[string]string `json:"resources"`
}

// BattleSpectateSpec 描述观战和回放能力。
type BattleSpectateSpec struct {
	EnableMonitor bool `json:"enable_monitor"`
	EnableReplay  bool `json:"enable_replay"`
}

// BattleLifecycleSpec 描述对抗赛轮次与生命周期配置。
type BattleLifecycleSpec struct {
	RoundDurationSeconds int  `json:"round_duration_seconds"`
	UpgradeWindowSeconds int  `json:"upgrade_window_seconds"`
	TotalRounds          int  `json:"total_rounds"`
	AutoCleanup          bool `json:"auto_cleanup"`
}

// BattleOrchestration 定义对抗赛的完整环境与赛制编排。
type BattleOrchestration struct {
	SharedChain   BattleSharedChainSpec   `json:"shared_chain"`
	Judge         BattleJudgeSpec         `json:"judge"`
	TeamWorkspace BattleTeamWorkspaceSpec `json:"team_workspace"`
	Spectate      BattleSpectateSpec      `json:"spectate"`
	Lifecycle     BattleLifecycleSpec     `json:"lifecycle"`
}

// Scan 实现 Scanner 接口。
func (o *BattleOrchestration) Scan(value interface{}) error {
	if value == nil {
		*o = BattleOrchestration{}
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return nil
	}

	return json.Unmarshal(bytes, o)
}

// Value 实现 Valuer 接口。
func (o BattleOrchestration) Value() (driver.Value, error) {
	return json.Marshal(o)
}

File: frontend/src/types/contestOrchestration.ts
/**
 * 比赛环境编排相关类型定义。
 * 这里专门承载解题赛与对抗赛的环境/赛制模型，避免和基础比赛类型混在一起。
 */

// 解题赛题目的主工作区配置。
export interface ChallengeWorkspaceSpec {
  image: string
  display_name?: string
  template?: string
  interaction_tools: string[]
  resources?: Record<string, string>
  init_scripts?: string[]
}

// 解题赛题目的附加服务端口。
export interface ChallengeServicePort {
  name: string
  port: number
  protocol: string
  expose_as?: string
}

// 解题赛题目的附加服务组件。
export interface ChallengeServiceSpec {
  key: string
  image: string
  purpose?: string
  description?: string
  ports?: ChallengeServicePort[]
  env?: Record<string, {} | undefined>
}

// 解题赛题目的拓扑说明。
export interface ChallengeTopologySpec {
  mode: string
  exposed_entries: string[]
  shared_network?: boolean
}

// 解题赛题目的 Fork 复现配置。
export interface ChallengeForkSpec {
  enabled: boolean
  chain?: string
  chain_id?: number
  label?: string
  rpc_url?: string
  block_number?: number
  target_tx_hash?: string
}

// 解题赛题目的场景说明。
export interface ChallengeScenarioSpec {
  contract_address?: string
  attack_goal?: string
  init_steps?: string[]
  solve_steps?: string[]
  defense_goal?: string
}

// 解题赛题目的生命周期配置。
export interface ChallengeLifecycleSpec {
  time_limit_minutes: number
  auto_destroy: boolean
  reuse_running_env: boolean
}

// 解题赛题目的验证配置。
export interface ChallengeValidationSpec {
  mode?: string
  description?: string
}

// 解题赛题目的正式环境编排模型。
export interface ChallengeOrchestration {
  mode: string
  needs_environment: boolean
  workspace: ChallengeWorkspaceSpec
  services: ChallengeServiceSpec[]
  topology: ChallengeTopologySpec
  fork: ChallengeForkSpec
  scenario: ChallengeScenarioSpec
  lifecycle: ChallengeLifecycleSpec
  validation: ChallengeValidationSpec
}

// 对抗赛共享链配置。
export interface BattleSharedChainSpec {
  image?: string
  chain_type?: string
  network_id?: number
  block_time?: number
  initial_balances?: Record<string, string>
  rpc_url?: string
}

// 对抗赛裁判与规则配置。
export interface BattleJudgeSpec {
  image?: string
  judge_contract?: string
  token_contract?: string
  strategy_interface?: string
  resource_model?: string
  scoring_model?: string
  score_weights?: Record<string, number>
  allowed_actions?: string[]
  forbidden_calls?: string[]
}

// 对抗赛队伍工作区模板。
export interface BattleTeamWorkspaceSpec {
  image?: string
  display_name?: string
  interaction_tools?: string[]
  resources?: Record<string, string>
}

// 对抗赛观战配置。
export interface BattleSpectateSpec {
  enable_monitor?: boolean
  enable_replay?: boolean
}

// 对抗赛生命周期配置。
export interface BattleLifecycleSpec {
  round_duration_seconds?: number
  upgrade_window_seconds?: number
  total_rounds?: number
  auto_cleanup?: boolean
}

// 对抗赛正式编排模型。
export interface BattleOrchestration {
  shared_chain: BattleSharedChainSpec
  judge: BattleJudgeSpec
  team_workspace: BattleTeamWorkspaceSpec
  spectate: BattleSpectateSpec
  lifecycle: BattleLifecycleSpec
}
